src/main/java/com/gildedrose/AgedBrie.java

```diff
@@ -1,27 +1,29 @@
 package com.gildedrose;
 
-public class AgedBrie {
+public class AgedBrie implements Ageable {
     private int qualityIncrementation;
+    private final Item item;
 
-    public AgedBrie() {
+    public AgedBrie(Item item) {
```

Comment:
When wrapping an object for composition, dependency injection via constructor argument (aka constructor injection) is a nice way to go. 👍

Reply:
idk why I originally thought that adding it to the method was a better decision since _at the time_ prior to this requirement my thought was that since the only method within these classes that would need information about `Item` is `.age()` so I felt that injecting it in the method was a better way to avoid a potential crash since the coupling of `Aged Brie` is tied to `Item` so if something in `Item` is broken then `Aged Brie` wont work but in our original implementation if `Item` was broken then only `.age()` wouldn't work.

As I'm typing this out, I also realize that this whole program is pretty much coupled to `Item` so if something in `Item` is broken then we prob wouldn't want the program running anyway.

But were we onto anything here? I have no concrete science to back up that argument but I feel like I'm either close to a broader idea with this or am I just talking out of my ass.

Reply:
> idk why I originally thought that adding it to the method was a better decision since at the time prior to this requirement my thought was that since the only method within these classes that would need information about Item is .age()

The idea of keeping information close to where it is being used is not a bad one. For example, another technique for GildedRose is for the `age` method to be declared `static` which means it does not (can not) have any other data dependencies than its function arguments (and fields declared `final` that have been initialized).  You'll use these type of pure functions more in functional programming contexts.  But that solution is not _better_ in a meaningful sense for what we're doing here.

Injecting the object being composed via the constructor makes it easier to treat all of these Item objects the same via polymorphism.  Once you have constructed one of your item types, they can all be treated the same. This will make working with collections of them easier.

You can achieve polymorphism via static functions with overloading. Clojure does this with "multimethods" for example.  But, it is just a different technique to accomplish the same goal. Polymorphism.

> As I'm typing this out, I also realize that this whole program is pretty much coupled to Item so if something in Item is broken then we prob wouldn't want the program running anyway.

That is correct. There's not much you can do about that coupling _within_ the same program.  You could wrap everything you do with an Item in a try/catch or something, but there's no real need to do that. And what would you do if Item raised an exception? You would probably need to crash and report what error occurred to the user, anyways.  You absolutely do not want a null or corrupted data when doing financial calculations. Unless you understand the domain and what the correct thing to do when recovering from an error.